### PR TITLE
Remove most of the metaprogramming in node.h in favor of listing all the members in the typed node structs.

### PR DESCRIPTION
**Pull request by zygoloid**
Split `node.h` into separate files for ID types (`id.h`) and for typed nodes (`typed_nodes.h`). The per-node-kind data is now specified as part of declaring the typed nodes, and is removed from the node kinds x-macros, which now simply enumerate the node kinds.

### CODE REVIEW

File: toolchain/sem_ir/node_kind.def
Comment (jonmeow) [inline toolchain/sem_ir/node_kind.def]: Maybe merge this into the main list? It doesn't feel special anymore, I don't think it needed to be first.

File: toolchain/sem_ir/node_kind.def
Comment (jonmeow) [inline toolchain/sem_ir/node_kind.def:20]:
```suggestion
// These have a 1:1 correspondence to entries in typed_nodes.h.
CARBON_SEM_IR_NODE_KIND(AddressOf)
```

Maybe add a comment in this file noting where to look for the rest?

File: toolchain/sem_ir/node_kind.h
Comment (jonmeow) [inline toolchain/sem_ir/node_kind.h:92]: Should probably have a comment explaining how this class is used, noting @josh11b's questions about that.

File: toolchain/sem_ir/typed_nodes.h
Comment (jonmeow) [inline toolchain/sem_ir/typed_nodes.h:487]: It might be worth moving these to the top, since otherwise they might be easily missed beneath all the actual typed nodes. I can understand though if you'd rather not; it only really makes sense if we don't have too much extra code like this.

File: toolchain/sem_ir/typed_nodes.h
Comment (josh11b) [inline toolchain/sem_ir/typed_nodes.h:40]: Please have a test that validates that you picked up the right enum value here.

File: toolchain/sem_ir/typed_nodes.h
Comment (josh11b) [inline toolchain/sem_ir/typed_nodes.h:43]: Please include a test that these two fields, when present, are present first and in the right order.

File: toolchain/sem_ir/typed_nodes.h
Comment (josh11b) [inline toolchain/sem_ir/typed_nodes.h:45]: Please have a test that each typed node can be converted from a Node and back.

File: toolchain/sem_ir/typed_nodes.h
Comment (josh11b) [inline toolchain/sem_ir/typed_nodes.h:487]: Document this technique in the toolchain doc.

File: toolchain/sem_ir/typed_nodes.h
Comment (josh11b) [inline toolchain/sem_ir/typed_nodes.h:13]: Document the static `Kind` member as well. It would be good to include a summary of what happens when you write that line. This includes associating the enum value of NodeKind::<Name> with this struct type, the ir_name string constant, and the terminator kind. Perhaps it would be good to have some breadcrumbs here that this file is included by node_kind.cpp, which uses this data to populate its tables?

File: toolchain/sem_ir/typed_nodes.h
Comment (josh11b) [inline toolchain/sem_ir/typed_nodes.h]:
```suggestion
// specified above when present. When converting to a `SemIR::Node`, they will become the
```

**Review by jonmeow — COMMENTED** [review]
A few minor comments, but generally I think this seems fine. I know @josh11b had mentioned some tests which would help. If it sticks like this I'm good with the look-and-feel, I do think it's an improvement. You don't need to wait for my approval -- I'm only really not approving because I'm sure others want to review.

File: toolchain/sem_ir/formatter.cpp
Comment (josh11b) [inline toolchain/sem_ir/formatter.cpp]: Would it be worth defining `Info::Arg1Field`?

File: toolchain/sem_ir/formatter.cpp
Comment (josh11b) [inline toolchain/sem_ir/formatter.cpp:670]: Perhaps we can put more functionality into `TypedNodeArgsInfo` so this code doesn't need to deal with `StructReflection` and tuples? Could we make it so you write `Info::GetField0(node)` instead of
```
auto tuple = StructReflection::AsTuple(node);
std::get<Info::Arg0Field>(tuple)
```

File: toolchain/sem_ir/node.h
Comment (josh11b) [inline toolchain/sem_ir/node.h:482]: Perhaps revive this comment with a pointer to `typed_nodes.h`?

File: toolchain/sem_ir/node.h
Comment (josh11b) [inline toolchain/sem_ir/node.h]:
```suggestion
// In addition, any kind-specific data can be accessed by casting
```

File: toolchain/sem_ir/node.h
Comment (josh11b) [inline toolchain/sem_ir/node.h]: Additional `Info` methods for getting the fields would be useful here too.

File: toolchain/sem_ir/node.h
Comment (josh11b) [inline toolchain/sem_ir/node.h:189]: Is the indentation weird here?

File: toolchain/sem_ir/typed_nodes.h
Comment (zygoloid) [reply toolchain/sem_ir/typed_nodes.h:487]: These to me feel like implementation details that most readers of this file won't be interested in, so I think I'd prefer to keep them here, or maybe move them to a different file?

File: toolchain/sem_ir/typed_nodes.h
Comment (zygoloid) [reply toolchain/sem_ir/typed_nodes.h:40]: Done.

File: toolchain/sem_ir/typed_nodes.h
Comment (zygoloid) [reply toolchain/sem_ir/typed_nodes.h:43]: Done.

File: toolchain/sem_ir/typed_nodes.h
Comment (zygoloid) [reply toolchain/sem_ir/typed_nodes.h:45]: Done.

File: toolchain/sem_ir/typed_nodes.h
Comment (zygoloid) [reply toolchain/sem_ir/typed_nodes.h:13]: Done.

File: toolchain/sem_ir/typed_nodes.h
Comment (zygoloid) [reply toolchain/sem_ir/typed_nodes.h]: Done.

File: toolchain/sem_ir/formatter.cpp
Comment (zygoloid) [reply toolchain/sem_ir/formatter.cpp]: Sure, done.

File: toolchain/sem_ir/formatter.cpp
Comment (zygoloid) [reply toolchain/sem_ir/formatter.cpp:670]: It feels a bit wasteful to me to convert the struct to a tuple once for each field you read, but I suppose that'll all get optimized away. Done.

File: toolchain/sem_ir/node.h
Comment (zygoloid) [reply toolchain/sem_ir/node.h:482]: The class comment already starts with: "that may be constructed from the specific kinds of node defined in `typed_nodes.h`", do you want more than that?

File: toolchain/sem_ir/node.h
Comment (zygoloid) [reply toolchain/sem_ir/node.h]: Done.

File: toolchain/sem_ir/node.h
Comment (zygoloid) [reply toolchain/sem_ir/node.h]: Done.

File: toolchain/sem_ir/node.h
Comment (zygoloid) [reply toolchain/sem_ir/node.h:189]: It looks OK to me, but github seems to not deal very well with the large deletions here in the side-by-side view.

File: toolchain/sem_ir/node_kind.def
Comment (zygoloid) [reply toolchain/sem_ir/node_kind.def]: Done.

File: toolchain/sem_ir/node_kind.def
Comment (zygoloid) [reply toolchain/sem_ir/node_kind.def:20]: Done.

File: toolchain/sem_ir/node_kind.h
Comment (zygoloid) [reply toolchain/sem_ir/node_kind.h:92]: Done.

File: toolchain/sem_ir/typed_nodes.h
Comment (zygoloid) [reply toolchain/sem_ir/typed_nodes.h:487]: Done.

Comment (josh11b) [reply toolchain/sem_ir/typed_nodes.h:487]: FYI, I agree this feels secondary to the typed nodes themselves. I'm not sure if I would consider it an implementation detail, because they are named by consumers of the header, but I would be fine with them at the end or in a different file. If it was in a different file, that file would naturally also contain `TypedNodeArgsInfo`.

File: toolchain/sem_ir/node_kind.h
Comment (josh11b) [inline toolchain/sem_ir/node_kind.h]: How about we make this into a sentence?

```suggestion
// A definition of a node kind is a value of the NodeKind itself plus ancillary data such as the name to use
```

File: toolchain/sem_ir/typed_nodes.h
Comment (josh11b) [inline toolchain/sem_ir/typed_nodes.h]: and something about how it associates this type with a particular value of the `NodeKind` enumerant?

File: toolchain/sem_ir/node_kind.h
Comment (zygoloid) [reply toolchain/sem_ir/node_kind.h]: Rephrased to include the "is a value of NodeKind" part.

(I found the suggested change as-is to make the comment harder for me to read -- generally I expect the first sentence in a class comment to be a noun phrase that could be preceded by "This class is...", in the same way that the first sentence in a class comment is missing its preceding "This function ...".)

File: toolchain/sem_ir/typed_nodes.h
Comment (zygoloid) [reply toolchain/sem_ir/typed_nodes.h]: Done.

**Review by josh11b — APPROVED** [review]

**Pull request merged**
